lib/ruby_indexer/lib/ruby_indexer/declaration_listener.rb

```diff
@@ -552,6 +582,34 @@ def handle_global_variable(node, loc)
       ))
     end
 
+    sig do
+      params(
+        node: T.any(
+          Prism::ClassVariableAndWriteNode,
+          Prism::ClassVariableOperatorWriteNode,
+          Prism::ClassVariableOrWriteNode,
+          Prism::ClassVariableTargetNode,
+          Prism::ClassVariableWriteNode,
+        ),
+        loc: Prism::Location,
+      ).void
+    end
+    def handle_class_variable(node, loc)
+      name = node.name.to_s
+      return if name == "@@"
+
+      comments = collect_comments(node)
+      owner = @owner_stack.last
+
+      @index.add(Entry::ClassVariable.new(
+        name,
+        @uri,
+        Location.from_prism_location(loc, @code_units_cache),
+        comments,
+        owner,
+      ))
```

Comment:
Ok, should we push two entries: one for the singleton class and another for the base class?

Reply:
Yes, correct. That will make the features properly find the variable both in a singleton context or in the attached class context, since it is available for both.

Reply:
I have updated the logic to add class variable entries to both the singleton class and the base class.
```ruby
    def handle_class_variable(node, loc)
      name = node.name.to_s
      return if name == "@@"

      comments = collect_comments(node)
      owner_stack = @owner_stack.reverse

      owner = owner_stack.shift
      @index.add(Entry::ClassVariable.new(
        name,
        @uri,
        Location.from_prism_location(loc, @code_units_cache),
        comments,
        owner,
      ))

      case owner
      when Entry::SingletonClass
        owner_stack.each do |owner|
          @index.add(Entry::ClassVariable.new(
            name,
            @uri,
            Location.from_prism_location(loc, @code_units_cache),
            comments,
            owner,
          ))

          break unless owner.is_a?(Entry::SingletonClass)
        end
      when Entry::Class
        @index.add(Entry::ClassVariable.new(
          name,
          @uri,
          Location.from_prism_location(loc, @code_units_cache),
          comments,
          @index.existing_or_new_singleton_class(owner.name),
        ))
      end
    end
```

If the owner is a class node, I add a class variable entry to both its class and its singleton class. If the node is a singleton class, I reverse the owner stack and continue adding the class variable to the owners in the stack, stopping once I reach the second owner that is not a singleton class. Bubbling up works as expected, but I am uncertain how to implement bubbling in.

[Screencast from 15-12-24 12:23:09 AM IST.webm](https://github.com/user-attachments/assets/da116fcb-267f-4e31-9d6a-75ff77d4b75b)

Reply:
I have updated the logic to add class variable entries to both the singleton class and the base class.
```ruby
    def handle_class_variable(node, loc)
      name = node.name.to_s
      return if name == "@@"

      comments = collect_comments(node)
      owner_stack = @owner_stack.reverse

      owner = owner_stack.shift
      @index.add(Entry::ClassVariable.new(
        name,
        @uri,
        Location.from_prism_location(loc, @code_units_cache),
        comments,
        owner,
      ))

      case owner
      when Entry::SingletonClass
        owner_stack.each do |owner|
          @index.add(Entry::ClassVariable.new(
            name,
            @uri,
            Location.from_prism_location(loc, @code_units_cache),
            comments,
            owner,
          ))

          break unless owner.is_a?(Entry::SingletonClass)
        end
      when Entry::Class
        @index.add(Entry::ClassVariable.new(
          name,
          @uri,
          Location.from_prism_location(loc, @code_units_cache),
          comments,
          @index.existing_or_new_singleton_class(owner.name),
        ))
      end
    end
```

If the owner is a class node, I add a class variable entry to both its class and its singleton class. If the node is a singleton class, I reverse the owner stack and continue adding the class variable to the owners in the stack, stopping once I reach the second owner that is not a singleton class. Bubbling up works as expected, but I am uncertain how to implement bubbling in.

[Screencast from 15-12-24 12:23:09 AM IST.webm](https://github.com/user-attachments/assets/da116fcb-267f-4e31-9d6a-75ff77d4b75b)

Reply:
I understand the confusion now. The changes in the type inferrer are assuming that class variables will behave exactly like instance variables, which is likely the source of the issue.

I think we may be able to do this:

1. In the declaration listener during indexing, only add an entry for the attached class (forget my previous comment about storing a copy in the singleton version)
2. Then in the type inferrer, when detecting the `self` type for class variables, we essentially always return whatever the attached class of the current context is, since the class variables will always be associated to them. Something like this

```ruby
class Foo
  # The inferred `self` here is `Foo` for class variables
  @@var = 123

  class << self
    # The inferred `self` here is also `Foo` for class variables
    @@other = 123
  end
end
```

I think that should lead to all of the features working as expected and it will simplify the declaration listener. Let me know if you find that this is not the right way.

Reply:
I have made the changes as per your suggestion.